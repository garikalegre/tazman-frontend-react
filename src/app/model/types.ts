import { ActionType, StateType } from 'typesafe-actions';
import { rootReducer } from '../store/root.reducer';
import * as AuthActions from '../duck/auth/auth.action';

const Actions = { ...AuthActions };

export type RootActions = ActionType<typeof Actions>;
export type RootState = StateType<typeof rootReducer>;
