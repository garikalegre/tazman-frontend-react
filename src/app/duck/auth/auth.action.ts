import { createStandardAction } from 'typesafe-actions';

import { UserLogin } from '../../../api/model/user.login';
import { FacebookLogin } from '../../../api/model/facebook.login';
import * as AuthTypes from './auth.types';

export const loginRequest = createStandardAction(AuthTypes.LOGIN_REQUEST)<
    UserLogin
>();
export const facebookLoginRequest = createStandardAction(
    AuthTypes.LOGIN_FACEBOOK_REQUEST
)<FacebookLogin>();
export const checkAuthRequest = createStandardAction(
    AuthTypes.CHECK_AUTH_REQUEST
)<undefined>();
export const loginFailure = createStandardAction(
    AuthTypes.LOGIN_ERROR
)<Error | null>();
export const loginSuccess = createStandardAction(AuthTypes.LOGIN_SUCCESS)<
    undefined
>();
