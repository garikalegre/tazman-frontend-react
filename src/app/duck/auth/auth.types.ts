export const LOGIN_REQUEST = '@auth/login/LOGIN_REQUEST';
export const LOGIN_SUCCESS = '@auth/login/LOGIN_SUCCESS';
export const LOGIN_ERROR = '@auth/login/LOGIN_ERROR';
export const LOGIN_FACEBOOK_REQUEST = '@auth/login/LOGIN_FACEBOOK_REQUEST';
export const CHECK_AUTH_REQUEST = '@auth/login/CHECK_AUTH_REQUEST';
