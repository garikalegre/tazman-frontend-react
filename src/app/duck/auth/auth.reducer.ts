import { Actions, State, initialState } from './auth.state';
import * as AuthTypes from './auth.types';

export const authReducer = (
    state: State = initialState,
    action: Actions
): State => {
    switch (action.type) {
        case AuthTypes.LOGIN_REQUEST:
        case AuthTypes.LOGIN_FACEBOOK_REQUEST:
        case AuthTypes.CHECK_AUTH_REQUEST:
            return { ...state, isAuth: false, isLoading: true, error: null };
        case AuthTypes.LOGIN_SUCCESS:
            return { ...state, isAuth: true, isLoading: false, error: null };
        case AuthTypes.LOGIN_ERROR:
            const error = action.payload;
            return { ...state, isAuth: false, isLoading: false, error };
        default:
            return state;
    }
};
