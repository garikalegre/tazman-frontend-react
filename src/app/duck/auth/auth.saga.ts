import { put, call, take, all, takeEvery } from 'redux-saga/effects';

import * as AuthActions from './auth.action';
import * as AuthTypes from './auth.types';
import { login } from '../../../api';
import {
    deleteFromStorage,
    getFromStorage,
    setToStorage
} from '../../../utils';

export function* loginSaga() {
    while (true) {
        try {
            let response;
            const action = yield take([
                AuthTypes.LOGIN_REQUEST,
                AuthTypes.LOGIN_FACEBOOK_REQUEST
            ]);
            response = yield call(login, action.payload);
            yield call(setToStorage, 'auth_token', response.token);
            yield put(AuthActions.loginSuccess());
        } catch (err) {
            yield call(deleteFromStorage, 'auth_token');
            yield put(AuthActions.loginFailure(err));
        }
    }
}

export function* checkAuthSaga() {
    const token = yield call(getFromStorage, 'auth_token');
    if (token) yield put(AuthActions.loginSuccess());
    else yield put(AuthActions.loginFailure(null));
}

export function* authSaga() {
    yield all([
        loginSaga(),
        takeEvery(AuthTypes.CHECK_AUTH_REQUEST, checkAuthSaga)
    ]);
}
