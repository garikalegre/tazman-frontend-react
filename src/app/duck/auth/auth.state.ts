import { ActionType } from 'typesafe-actions';
import * as AuthActions from './auth.action';

export type State = {
    isLoading: boolean;
    isAuth: boolean;
    error: Error | null;
};

export type Actions = ActionType<typeof AuthActions>;

export const initialState: State = {
    isLoading: true,
    isAuth: false,
    error: null
};
