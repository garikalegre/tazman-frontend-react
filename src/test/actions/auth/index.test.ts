import * as AuthActions from '../../../app/duck/auth/auth.action';
import * as AuthTypes from '../../../app/duck/auth/auth.types';

describe('Auth actions', () => {
    it(`creates a ${AuthTypes.LOGIN_REQUEST} action`, () => {
        expect(
            AuthActions.loginRequest({ email: 'email', password: 'password' })
        ).toMatchSnapshot();
    });
    it(`creates a ${AuthTypes.CHECK_AUTH_REQUEST} action`, () => {
        expect(AuthActions.checkAuthRequest()).toMatchSnapshot();
    });
    it(`creates a ${AuthTypes.LOGIN_FACEBOOK_REQUEST} action`, () => {
        expect(
            AuthActions.facebookLoginRequest({
                facebookId: 'facebookId',
                token: 'token'
            })
        ).toMatchSnapshot();
    });
    it(`creates a ${AuthTypes.LOGIN_SUCCESS} action`, () => {
        expect(AuthActions.loginSuccess()).toMatchSnapshot();
    });
    it(`creates a ${AuthTypes.LOGIN_ERROR} action`, () => {
        expect(AuthActions.loginFailure(new Error())).toMatchSnapshot();
    });
});
