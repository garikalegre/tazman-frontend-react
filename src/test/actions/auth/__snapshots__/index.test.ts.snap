// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth actions creates a @auth/login/CHECK_AUTH_REQUEST action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "@auth/login/CHECK_AUTH_REQUEST",
}
`;

exports[`Auth actions creates a @auth/login/LOGIN_ERROR action 1`] = `
Object {
  "meta": undefined,
  "payload": [Error],
  "type": "@auth/login/LOGIN_ERROR",
}
`;

exports[`Auth actions creates a @auth/login/LOGIN_FACEBOOK_REQUEST action 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "facebookId": "facebookId",
    "token": "token",
  },
  "type": "@auth/login/LOGIN_FACEBOOK_REQUEST",
}
`;

exports[`Auth actions creates a @auth/login/LOGIN_REQUEST action 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "email": "email",
    "password": "password",
  },
  "type": "@auth/login/LOGIN_REQUEST",
}
`;

exports[`Auth actions creates a @auth/login/LOGIN_SUCCESS action 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "@auth/login/LOGIN_SUCCESS",
}
`;
